From bf70301c2f9cba642fd663dc8d99772d90a757c2 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <grinberg@compulab.co.il>
Date: Mon, 4 Mar 2013 17:08:53 +0200
Subject: ARM: i.MX6: cm-fx6: add SanDisk iSSD support

Add support for the on-board cm-fx6 SanDisk iSSD.

Signed-off-by: Igor Grinberg <grinberg@compulab.co.il>
---
 arch/arm/mach-mx6/board-cm-fx6-q.h |   12 +++++++++
 arch/arm/mach-mx6/board-cm-fx6.c   |   49 ++++++++++++++++++++++++++++++++++--
 2 files changed, 59 insertions(+), 2 deletions(-)

diff --git a/arch/arm/mach-mx6/board-cm-fx6-q.h b/arch/arm/mach-mx6/board-cm-fx6-q.h
index dc1ea15..3c06b6b 100644
--- a/arch/arm/mach-mx6/board-cm-fx6-q.h
+++ b/arch/arm/mach-mx6/board-cm-fx6-q.h
@@ -143,6 +143,18 @@ static iomux_v3_cfg_t cm_fx6_q_common_pads[] = {
 
 	/* USBOTG ID pin */
 	MX6Q_PAD_GPIO_1__USBOTG_ID,
+
+	/* SATA PWR */
+	MX6Q_PAD_ENET_TX_EN__GPIO_1_28,	/* SATA_PWREN */
+	MX6Q_PAD_EIM_A22__GPIO_2_16,	/* SATA_LDO_EN */
+	MX6Q_PAD_EIM_D20__GPIO_3_20,	/* SATA_nSTANDBY1 */
+	MX6Q_PAD_EIM_A25__GPIO_5_2,	/* SATA_nSTANDBY2 */
+	/* SATA CTRL */
+	MX6Q_PAD_ENET_TXD0__GPIO_1_30,	/* SATA_VDDC_CTRL */
+	MX6Q_PAD_EIM_D23__GPIO_3_23,	/* SATA_PHY_SLP */
+	MX6Q_PAD_EIM_D29__GPIO_3_29,	/* SATA_STBY_REQ */
+	MX6Q_PAD_EIM_A23__GPIO_6_6,	/* SATA_nRSTDLY */
+	MX6Q_PAD_EIM_BCLK__GPIO_6_31,	/* SATA_PWLOSS_INT */
 };
 
 static iomux_v3_cfg_t cm_fx6_q_spdif_pads[] = {
diff --git a/arch/arm/mach-mx6/board-cm-fx6.c b/arch/arm/mach-mx6/board-cm-fx6.c
index 20b832e..a0ebe97 100644
--- a/arch/arm/mach-mx6/board-cm-fx6.c
+++ b/arch/arm/mach-mx6/board-cm-fx6.c
@@ -81,13 +81,22 @@
 #include "board-cm-fx6-q.h"
 #include "board-cm-fx6-dl.h"
 
-/* GPIO PIN, sort by PORT/BIT */
+/* GPIO PIN, sort by board/PORT/BIT */
+#define CM_FX6_SATA_PWREN		IMX_GPIO_NR(1, 28)
+#define CM_FX6_SATA_VDDC_CTRL		IMX_GPIO_NR(1, 30)
 #define CM_FX6_ADS7846_PENDOWN		IMX_GPIO_NR(2, 15)
+#define CM_FX6_SATA_LDO_EN		IMX_GPIO_NR(2, 16)
 #define CM_FX6_ECSPI1_CS0		IMX_GPIO_NR(2, 30)
 #define CM_FX6_GREEN_LED		IMX_GPIO_NR(2, 31)
 #define CM_FX6_ECSPI1_CS1		IMX_GPIO_NR(3, 19)
+#define CM_FX6_SATA_nSTANDBY1		IMX_GPIO_NR(3, 20)
+#define CM_FX6_SATA_PHY_SLP		IMX_GPIO_NR(3, 23)
 #define CM_FX6_ECSPI2_CS2		IMX_GPIO_NR(3, 24)
 #define CM_FX6_ECSPI2_CS3		IMX_GPIO_NR(3, 25)
+#define CM_FX6_SATA_STBY_REQ		IMX_GPIO_NR(3, 29)
+#define CM_FX6_SATA_nSTANDBY2		IMX_GPIO_NR(5, 2)
+#define CM_FX6_SATA_nRSTDLY		IMX_GPIO_NR(6, 6)
+#define CM_FX6_SATA_PWLOSS_INT		IMX_GPIO_NR(6, 31)
 #define CM_FX6_USB_HUB_RST		IMX_GPIO_NR(7, 8)
 
 #define SB_FX6_HX8520_PENDOWN		IMX_GPIO_NR(1, 4)
@@ -644,6 +653,26 @@ static void __init cm_fx6_init_usb(void)
 	defined(CONFIG_SATA_AHCI_PLATFORM_MODULE)
 static struct clk *sata_clk;
 
+static struct gpio sata_issd_gpios[] = {
+	/* The order of the GPIOs in the array is important! */
+	{ CM_FX6_SATA_PHY_SLP,	 GPIOF_OUT_INIT_LOW,	"sata phy slp" },
+	{ CM_FX6_SATA_nRSTDLY,	 GPIOF_OUT_INIT_LOW,	"sata nrst" },
+	{ CM_FX6_SATA_PWREN,	 GPIOF_OUT_INIT_LOW,	"sata pwren" },
+	{ CM_FX6_SATA_nSTANDBY1, GPIOF_OUT_INIT_LOW,	"sata nstndby1" },
+	{ CM_FX6_SATA_nSTANDBY2, GPIOF_OUT_INIT_LOW,	"sata nstndby2" },
+	{ CM_FX6_SATA_LDO_EN,	 GPIOF_OUT_INIT_LOW,	"sata ldo en" },
+};
+
+static void cm_fx6_sata_power_on(bool on)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(sata_issd_gpios); i++) {
+		gpio_set_value(sata_issd_gpios[i].gpio, on ? 1 : 0);
+		udelay(100);
+	}
+}
+
 /* HW Initialization, if return 0, initialization is successful. */
 static int cm_fx6_sata_init(struct device *dev, void __iomem *addr)
 {
@@ -651,11 +680,21 @@ static int cm_fx6_sata_init(struct device *dev, void __iomem *addr)
 	int err;
 	struct clk *clk;
 
+	/* SATA PWR GPIOs */
+	err = gpio_request_array(sata_issd_gpios, ARRAY_SIZE(sata_issd_gpios));
+	if (err) {
+		dev_err(dev, "sata power GPIOs request failed: %d\n", err);
+		return err;
+	}
+	udelay(100);
+
+	cm_fx6_sata_power_on(true);
+
 	sata_clk = clk_get(dev, "imx_sata_clk");
 	if (IS_ERR(sata_clk)) {
 		err = PTR_ERR(sata_clk);
 		dev_err(dev, "no sata clock err: %d\n", err);
-		return err;
+		goto free_gpios;
 	}
 
 	err = clk_enable(sata_clk);
@@ -704,6 +743,9 @@ release_sata_clk:
 	clk_disable(sata_clk);
 put_sata_clk:
 	clk_put(sata_clk);
+free_gpios:
+	cm_fx6_sata_power_on(false);
+	gpio_free_array(sata_issd_gpios, ARRAY_SIZE(sata_issd_gpios));
 
 	return err;
 }
@@ -712,6 +754,9 @@ static void cm_fx6_sata_exit(struct device *dev)
 {
 	clk_disable(sata_clk);
 	clk_put(sata_clk);
+
+	cm_fx6_sata_power_on(false);
+	gpio_free_array(sata_issd_gpios, ARRAY_SIZE(sata_issd_gpios));
 }
 
 static struct ahci_platform_data cm_fx6_sata_pdata = {
-- 
1.7.9.5

